python/examples/llama/test_llama3.py

```diff
@@ -0,0 +1,1861 @@
+                self.head_dim,
+            )
+        )
+        self.cache_v = torch.zeros(
+            (
+                args.max_batch_size,
+                args.max_seq_len,
+                self.n_kv_heads,
+                self.head_dim,
+            )
+        )
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        start_pos: int,
+        freqs_cis: torch.Tensor,
+        mask: Optional[torch.Tensor],
+    ):
+        bsz, seqlen, _ = x.shape
+        xq, xk, xv = self.wq(x), self.wk(x), self.wv(x)
+
+        xq = xq.view(bsz, seqlen, self.n_heads, self.head_dim)
+        xk = xk.view(bsz, seqlen, self.n_kv_heads, self.head_dim)
+        xv = xv.view(bsz, seqlen, self.n_kv_heads, self.head_dim)
+
+        xq, xk = rotary_emb_ref(xq, xk, freqs_cis=freqs_cis)
+
+        self.cache_k = self.cache_k.to(xq)
+        self.cache_v = self.cache_v.to(xq)
+
+        self.cache_k[:bsz, start_pos : start_pos + seqlen] = xk
+        self.cache_v[:bsz, start_pos : start_pos + seqlen] = xv
+
+        keys = self.cache_k[:bsz, : start_pos + seqlen]
+        values = self.cache_v[:bsz, : start_pos + seqlen]
+
+        # repeat k/v heads if n_kv_heads < n_heads
+        keys = repeat_kv(
+            keys, self.n_rep
+        )  # (bs, cache_len + seqlen, n_heads, head_dim)
+        values = repeat_kv(
+            values, self.n_rep
+        )  # (bs, cache_len + seqlen, n_heads, head_dim)
+
+        xq = xq.transpose(1, 2)  # (bs, n_heads, seqlen, head_dim)
+        keys = keys.transpose(1, 2)  # (bs, n_heads, cache_len + seqlen, head_dim)
+        values = values.transpose(1, 2)  # (bs, n_heads, cache_len + seqlen, head_dim)
+        scores = torch.matmul(xq, keys.transpose(2, 3)) / pymath.sqrt(self.head_dim)
+        if mask is not None:
+            scores = scores + mask  # (bs, n_heads, seqlen, cache_len + seqlen)
+        scores = torch.nn.functional.softmax(scores.float(), dim=-1).type_as(xq)
+        output = torch.matmul(scores, values)  # (bs, n_heads, seqlen, head_dim)
+        output = output.transpose(1, 2).contiguous().view(bsz, seqlen, -1)
+        return self.wo(output)
+
+
+references = {
+    get_add: torch.add,
+    get_mul: torch.mul,
+    get_matmul: torch.matmul,
+    get_rsqrt: torch.rsqrt,
+    get_sqr: torch.square,
+    get_mean: lambda x: torch.mean(x, dim=-1, keepdim=True),
+    get_silu: lambda x: torch.nn.functional.silu(x),
+    get_softmax: lambda x: torch.softmax(x, dim=-1),
+    get_polar: torch.polar,
+    get_triu: torch.triu,
+    get_outer: torch.outer,
+    get_linear: torch.nn.functional.linear,
+    get_repeat_kv: repeat_kv,
+    get_l2_norm: lambda x, eps: x
+    * torch.rsqrt(torch.mean(x.pow(2), dim=-1, keepdim=True) + eps),
+    get_rotary_emb: rotary_emb_ref,
+}
+
+
+# TODO: torch_dtype_to_mlir_type
+def to_ir_type(type_str, ctx):
+    if type_str == "f32":
+        return ir.F32Type.get(context=ctx)
+    elif type_str == "f64":
+        return ir.F64Type.get(context=ctx)
```

Comment:
```suggestion
def to_ir_type(type_str):
    if type_str == "f32":
        return ir.F32Type.get()
    elif type_str == "f64":
        return ir.F64Type.get()
```

In effect, these `.get()` methods are doing a `ir.Context.current` under the hood when you don't pass a context explicitly (just like the Op builders).